### PR TITLE
[Fix] Address remaining issues of supporting MiniCPMV

### PR DESCRIPTION
**Pull request by mickqian**
## Motivation

Address remaining issues of #2785 

## Modifications

1. Update document of implementing a new vision-llm
2. Add some test for comparing logits output of SGLang and HF
3. Code cleanup

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [x] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html).

### CODE REVIEW

**Comment by zhaochenyang20** [comment]
Great work! Once ready, please ask us to review.

**Comment by merrymercy** [comment]
also remove these vllm dependency https://github.com/sgl-project/sglang/blob/2584f6d94487645c48696762194457f6296c5ea7/python/sglang/srt/layers/attention/vision.py#L8-L9

File: python/sglang/srt/layers/attention/vision.py [hunk 1]
```diff
@@ -160,45 +165,62 @@ def forward(
...
-            q,
-            k,
-            v,
-            output,
-            cu_seqlens.cuda(),
-            seq_lens,
-            max_seqlen,
-            is_causal=False,
-        )
+        if self.use_context_forward:
```
Comment (mickqian) [inline python/sglang/srt/layers/attention/vision.py]: `context_attention_fwd` generates different result with hf implementation: `SiglipAttention`, probably because:
1.`SiglipAttention` performs softmax in float32:
```python
attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(query_states.dtype)
```
2. `SiglipAttention` performs full-sequence + mask, whereas `context_attention_fwd` skips padding tokens, leaving 0s in the attention weights.

**Comment by zhaochenyang20** [comment]
Cool. I will ask @yizhang2077 to help~

File: test/srt/test_vision_llm.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        with torch.no_grad():
+            input_ids = inputs["input_ids"].to(self.device).flatten()
+            image_inputs = model_runner.model._parse_and_validate_inputs(
+                input_ids=inputs["input_ids"].to(self.device).flatten(),
+                **{
+                    "pixel_values": [inputs["pixel_values"]],
+                    "tgt_sizes": [inputs["tgt_sizes"]],
+                    "im_start_id": [self.tokenizer.im_start_id],
+                    "im_end_id": [self.tokenizer.im_end_id],
+                    "slice_start_id": [self.tokenizer.slice_start_id],
```
Comment (yizhang2077) [inline test/srt/test_vision_llm.py:198]: Cool! I think maybe this test can be more general so that it can also be used by qwen2vl or more VLM?

File: python/sglang/srt/layers/attention/vision.py (hunk 1 above)
Comment (yizhang2077) [reply python/sglang/srt/layers/attention/vision.py]: Does qwen2vl model have the same question? If so, I think maybe we can set qwen2vl  use_context_forward to False or even remove use_context_forward branch.

File: test/srt/test_vision_openai_server.py [hunk 3]
```diff
@@ -457,6 +457,8 @@ def setUpClass(cls):
                 "--trust-remote-code",
                 "--chat-template",
                 "minicpmv",
+                "--max-total-tokens",
```
Comment (yizhang2077) [inline test/srt/test_vision_openai_server.py]: Why do we need add args here?

File: python/sglang/srt/layers/attention/vision.py (hunk 1 above)
Comment (mickqian) [reply python/sglang/srt/layers/attention/vision.py]: There are three optional self_attns supported by qwen2vl:
```python
QWEN2_VL_VISION_ATTENTION_CLASSES = {
    "eager": VisionAttention,
    "flash_attention_2": VisionFlashAttention2,
    "sdpa": VisionSdpaAttention,
}
```
where:
1. softmax:
   - eager adopts full-precision softmax
   - flash_attention and sdpa adopts half-precision softmax
2. attention:
   - flash_attention adopts sglang-like attention
   - eager and sdpa performs full-sequence attention

And qwen2 defaults to `sdpa`, so is it feasible to still keep the `use_context_forward` branch but set `False` for qwen2vl?

File: test/srt/test_vision_openai_server.py (hunk 3 above)
Comment (mickqian) [reply test/srt/test_vision_openai_server.py]: It is due to the OOM from full-sequence vision attention, specifically the full-precision softmax calculation. Is there a better way to address it?

File: python/sglang/srt/layers/attention/vision.py (hunk 1 above)
Comment (yizhang2077) [reply python/sglang/srt/layers/attention/vision.py]: I think it is ok

File: python/sglang/srt/models/minicpmv.py [hunk 4]
```diff
@@ -1044,13 +1100,12 @@ def get_vision_hidden_states(
         )
         for i in range(B):
             patch_attn_mask[i, 0, : tgt_sizes[i][0] * tgt_sizes[i][1]] = True
```
Comment (yizhang2077) [inline python/sglang/srt/models/minicpmv.py]: can you vectorize code and useless python for loop? reference [here](https://github.com/sgl-project/sglang/pull/2365/files#r1872859620)

File: python/sglang/srt/layers/attention/vision.py (hunk 1 above)
Comment (yizhang2077) [reply python/sglang/srt/layers/attention/vision.py]: can we directly use pytorch `F.scaled_dot_product_attention` for sdpa case?

File: test/srt/test_vision_llm.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+    def setUpClass(cls):
+        cls.image_url = "https://github.com/sgl-project/sglang/blob/main/test/lang/example_image.png?raw=true"
+        cls.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        cls.model_path = ""
+        cls.chat_template = ""
+        cls.processor = ""
+        response = requests.get(cls.image_url)
+        cls.main_image = Image.open(BytesIO(response.content))
+
+    def compare_outputs(self, sglang_output: torch.Tensor, hf_output: torch.Tensor):
```
Comment (mickqian) [inline test/srt/test_vision_llm.py:35]: Copied from #2365, hope that's ok @jakep-allenai

**Comment by zhaochenyang20** [comment]
@mickqian mick, great work. So we are ended and just need to past the review? Thanks!

File: test/srt/test_vision_llm.py (hunk 5 above)
Comment (jakep-allenai) [reply test/srt/test_vision_llm.py:35]: 👍

**Comment by yizhang2077** [comment]
@mickqian there are still some failed cases, but it seems they are not related to your pr?

**Comment by zhaochenyang20** [comment]
> @mickqian there are still some failed cases, but it seems they are not related to your pr?

Yeah. CI is wrong possibly due to kernel or largely refactor.

**Comment by yizhang2077** [comment]
> Yeah. CI is wrong possibly due to kernel or largely refactor.

Since `test_vision_*.py`  has been all passed,  I think we can approve this pr

**Review by yizhang2077 — APPROVED** [review]

File: test/srt/test_vision_openai_server.py [hunk 6]
```diff
@@ -381,6 +383,7 @@ def setUpClass(cls):
                 "--context-length",
                 "300",
                 "--mem-fraction-static=0.80",
+                "--disable-cuda-graph",
```
Comment (zhyncs) [inline test/srt/test_vision_openai_server.py]: QQ Why disable CUDA Graph?

Comment (yizhang2077) [reply test/srt/test_vision_openai_server.py]: Oh, I miss it

Comment (mickqian) [reply test/srt/test_vision_openai_server.py]: fixed
